### DISCUSSION TITLE
Can't save Auth

### DISCUSSION THREAD

**Discussion by YusufSuleman** [pocketbase/pocketbase#1965] [Q&A]
Hello, I am using SolidJS and PB, to create a blog site. I created a successful login page, But as soon as I move to another page, all the auth data in Local Storage is gone.
Here is the login code:
```tsx
const [fields, setFields] = createStore();
const pb = new PocketBase('http://127.0.0.1:8090');

//@ts-expect-error
async function login(e) {
    e.preventDefault();

    await pb.collection('users').authWithPassword(fields.email, fields.password);

    console.log(pb.authStore.isValid);
    console.log(pb.authStore.token);
    console.log(pb.authStore.model?.id);
}
```
They get set to 
```json
{"token":"","model":null}
```

**Comment by ganigeorgiev** [comment]
There is not enough details to help you debug it.

> all the auth data in Local Storage is gone.

What do you mean by "gone"?

If you are in a browser env `window.localStorage` is cleared on page reload only if you browser doesn't allow `localStorage` access, after closing a private tab/window OR something in your code is triggering `pb.authStore.clear()`.

My guess is that this is related to your other questions about SolidJS (https://github.com/pocketbase/pocketbase/discussions/1946, https://github.com/pocketbase/pocketbase/discussions/1941) but the code sample lack context to understand what is happening.

**Reply by YusufSuleman** [reply]
I can't find any thing related to clear, here is the whole code if that helps:
```tsx
import { createEffect, createResource } from "solid-js";
import { createStore } from "solid-js/store";
import PocketBase from 'pocketbase';

function Login() {
    const [fields, setFields] = createStore();
    const pb = new PocketBase('http://127.0.0.1:8090');

    //@ts-expect-error
    async function login(e) {
        e.preventDefault();
    
        await pb.collection('users').authWithPassword(fields.email, fields.password);
    
        console.log(pb.authStore.isValid);
        console.log(pb.authStore.token);
        console.log(pb.authStore.model?.id);
    }
    return(
        <div class="content-body">
        <div class="content-body register">
            <form onSubmit={login}>
                Email: <br />
                <input type="text" name="email" placeholder="Email" onInput={(e) => {setFields("email", e.target.value)}} /> <br />
                Password <br />
                <input type="password" name="password" placeholder="Password" minlength="10" onInput={(e) => {setFields("password", e.target.value)}}  /> <br />
                {/* Password Confirm: <br /> */}
                {/* <input type="password" name="passwordConfirm" placeholder="Password Confirm" minlength="10" onInput={(e) => {setPasswordConfirm(e.currentTarget.value)}} /> <br /> */}
                <button type="submit">Login</button>
            </form>
        </div>
        </div>
    )
}

export default Login
```

**Reply by ganigeorgiev** [reply]
@YusufSuleman Once again, there is not enough details to help you debug it. The above code doesn't show how you navigate between different pages and where do you check the auth state. The login function as in https://github.com/pocketbase/pocketbase/discussions/1946#discussioncomment-5163792, works fine for me using the [SolidJS playground](https://playground.solidjs.com/).

Could you provide a minimal reproducible repo (or a playground link)?

**Reply by YusufSuleman** [reply]
I don't check the auth state, while I was trying to get the state in the Navbar, it said that I am not verified, and after some more debugging, I found out that it was not storing local storage correctly in DevTools, I use `<a href="/register" class="item">Register</a>` for the navigation, here is the whole navbar code with comments:
```tsx
import PocketBase from 'pocketbase';
import { createResource, Show } from 'solid-js';

function Navbar() {
    const pb = new PocketBase('http://127.0.0.1:8090');
    const pocketbaseAuth:any = localStorage.getItem('pocketbase_auth');
    console.log(pocketbaseAuth)
    const authData = createResource(async () => { await pb.collection('users').authWithPassword(
        JSON.parse(pocketbaseAuth).model.email,
        JSON.parse(pocketbaseAuth).model.password,
    );})//Ok for some reason, auth is empty
    authData[0]()
    console.log("veri", pb.authStore.isValid) //False

    return (
        <div class="navbar">
            <div class="main">
                <a href="/" class="item active">Home</a>
                <a href="/blog" class="item">Blog</a>
                <a href="" class="item">Epics</a>

            </div>

            <div class="secondary">
                <Show
                    when={pb.authStore.isValid}
                >
                    <a href="/login" class="item">Users</a>
                </Show>
                <Show
                    when={!pb.authStore.isValid}
                >
                    <a href="/register" class="item">Register</a>
                    <a href="/login" class="item">Login</a>
                </Show>
                <a href="/about" class="item">About</a>
            </div>
        </div>
    )
}

export default Navbar;
```

**Reply by ganigeorgiev** [reply]
I don't understand what you are doing.

The plain password is not stored in `authStore.model` and you don't have to reauthenticate calling `authWithPassword`. If you want to refresh or check if the user is authenticated, you can call directly `pb.collection('users').authRefresh()`.

Additionally, `authData[0]()` may not have completed yet since it is async so the console log right after it may not return what you expect.

In any case, I don't think this is related to PocketBase or the SDK. As mentioned previously, I'm not familiar with SolidJS and I'm not sure what to recommend or suggest. You may want to ask in the SolidJS official support channel for further help.

**Reply by ganigeorgiev** [reply]
Additionally, please keep in mind that `pb.authStore.isValid` checks are not "reactive" so I'm not sure how the checks `<Show when={!pb.authStore.isValid}>` works in this case. As far as I understand from the documentation, SolidJS uses ["signals"](https://www.solidjs.com/docs/latest#createsignal) for reactivity, but once again, I have no experience with SolidJS and I'm not sure if this is the recommended approach.

**Reply by YusufSuleman** [reply]
Now it gives this in the console of the browser:
```
ClientResponseError.ts:13 Uncaught (in promise) ClientResponseError 401: The request requires valid record authorization token to be set.
    at new ClientResponseError (ClientResponseError.ts:13:9)
    at Client.ts:277:23
```
Same error in the PB logs

**Reply by ganigeorgiev** [reply]
I'm really confused. Please provide a minimal reproducible repo.

**Reply by YusufSuleman** [reply]
Here is the minified version of it containing only the code that is needed:
https://github.com/YusufSuleman/mmr
Login with example@example.com and example123 for PB and the same with for the actual site login. It is filled with gibberish example posts. Run `npm i` and `npm run dev`

I orignally started to create a new solid repo but too many errors and I just gave you a simplified version with personal info removed.

**Reply by ganigeorgiev** [reply]
@YusufSuleman Thank you for the repo but I still don't understand what you are trying to do. The repo also seems to be missing a dependency (`@solidjs/router`).

Couple things:

1. In a browser env you don't have to create every time `new PocketBase('http://127.0.0.1:8090')` when you want to make a call to the API. You can move it in its own file and then import it.

2. `pb.authStore.isValid` or anything else from the SDK client is not "reactive" and will not magically work with the SolidJS reactivity. Please read carefully the [SolidJS documentation and more specifically "signals"](https://www.solidjs.com/docs/latest). **I cannot provide general support or recommendations for SolidJS since I have no experience with it.**

3. I don't understand what are trying to do in the [Navbar component](https://github.com/YusufSuleman/mmr/blob/main/src/Navbar.tsx).
You don't have to manually read and parse the localStorage. On PocketBase SDK initialization `pb.authStore` will be automatically populated with the `localStorage` state (if any). `.authRefresh()` can be called if you want to validate the loaded local storage token and to get up-to-date user data. If all you want to do is to check whether the user was "logged in", then "pb.authStore.isValid" should suffice, but once again this is NOT reactive and will not trigger an update to your Solid component. **You have to use stores, signals, or any other SolidJS reactivity primitive.**

Or in other words:

1. If your app is entirely client-side, create a new js file (eg. `pb.js`):
   ```js
   // src/pb.js
   import PocketBase from 'pocketbase';

   const pb = new PocketBase('http://127.0.0.1:8090');
   
   export default pb;
   ```
 
2. Instead of `import PocketBase from 'pocketbase'`, import directly the above file:
   ```js
   import pb from "./pb.js"

   ...

   function Navbar() {
    console.log(pb.authStore.isValid)
    ...
   ```

3. Login and then refresh the page and you'll see that the above console.log will output `true`, but once again, if you need reactivity in the component you have to use the SolidJS reactivity primitives.

**Reply by ganigeorgiev** [reply]
One minor thing that I forgot to mention is that if you want to listen to `pb.authStore` changes you can register listener(s) via the `pb.authStore.onChange()`:

```js
// triggered every time on store change
pb.authStore.onChange((token, model) => {
    // eventually trigger an effect/signal/etc.
    console.log(token, model)
}, true); // true or false indicates whether also to fire the callback right after the registration
```

**Reply by YusufSuleman** [reply]
Made PB into a seperate file changed for both Navbar and Login. Still the same errors the onChange prints null for the Navbar but correct data for Login. Maybe the Solid Router is causing this?

**Reply by YusufSuleman** [reply]
@ganigeorgiev ?

**Reply by ganigeorgiev** [reply]
@YusufSuleman  I'm not able to reproduce it and nothing has changed since my last comment in https://github.com/pocketbase/pocketbase/discussions/1965#discussioncomment-5202289.
